OpenRA.Mods.Common/Widgets/Logic/Settings/SettingsLogic.cs

```diff
@@ -0,0 +1,160 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2020 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using OpenRA.Graphics;
+using OpenRA.Mods.Common.Traits;
+using OpenRA.Primitives;
+using OpenRA.Support;
+using OpenRA.Widgets;
+
+namespace OpenRA.Mods.Common.Widgets.Logic
+{
+	public class SettingsLogic : ChromeLogic
+	{
+		readonly Dictionary<string, Action> leavePanelActions = new Dictionary<string, Action>();
+		readonly Dictionary<string, Action> resetPanelActions = new Dictionary<string, Action>();
+
+		readonly Widget panelContainer, tabContainer;
+		readonly ButtonWidget tabTemplate;
+		readonly int2 buttonStride;
+		readonly List<ButtonWidget> buttons = new List<ButtonWidget>();
+		readonly List<Tuple<string, string>> panels = new List<Tuple<string, string>>();
+		string activePanel = "";
+
+		public static bool NeedsRestart = false;
+
+		static SettingsLogic() { }
+
+		[ObjectCreator.UseCtor]
+		public SettingsLogic(Widget widget, Action onExit, WorldRenderer worldRenderer, Dictionary<string, MiniYaml> logicArgs)
+		{
+			panelContainer = widget.Get("PANEL_CONTAINER");
+			var panelTemplate = panelContainer.Get<ContainerWidget>("PANEL_TEMPLATE");
+			panelContainer.RemoveChild(panelTemplate);
+
+			tabContainer = widget.Get("TAB_CONTAINER");
+			tabTemplate = tabContainer.Get<ButtonWidget>("BUTTON_TEMPLATE");
+			tabContainer.RemoveChild(tabTemplate);
+
+			if (logicArgs.TryGetValue("ButtonStride", out var buttonStrideNode))
+				buttonStride = FieldLoader.GetValue<int2>("ButtonStride", buttonStrideNode.Value);
+
+			if (logicArgs.TryGetValue("Panels", out var settingsPanels))
+			{
+				foreach (var settingsPanel in settingsPanels.Nodes)
+				{
+					var panelID = settingsPanel.Value.Nodes.FirstOrDefault(n => n.Key == "Id").Value.Value;
+					var label = settingsPanel.Value.Nodes.FirstOrDefault(n => n.Key == "Label").Value.Value;
+					var container = panelTemplate.Clone() as ContainerWidget;
+					container.Id = panelID;
+					panelContainer.AddChild(container);
+
+					panels.Add(Tuple.Create(panelID, label));
+
+					Game.LoadWidget(worldRenderer.World, panelID, container, new WidgetArgs()
+					{
+						{ "settingsLogic", this },
```

Comment:
I'm not sure if passing the whole instance here is a good idea. Only access to the method `RegisterSettingsPanel` is necessary but I wasn't sure how to do it.

Reply:
We could pass `{ "registerPanel", (Action<string, string, Func<Widget, Action>, Func<Widget, Action>>)RegisterSettingsPanel },` instead and invoke it directly in the panels?

Reply:
Yes, that works.